### PR TITLE
Doc2Vec Model and Evaluation

### PR DESCRIPTION
**Pull request by chelsealiannegomez**
## Description
I created a Doc2Vec model using TensorFlow and NumPy. This model uses the context from the articles to be able to predict a target word given its context words and the document ID. This allows documents to have unique embeddings that represent the overall context of their content. Then, given a query that is converted into embeddings using the trained word embeddings, it can compute the cosine similarity between the embeddings and return a ranked list of most similar articles to the query. 

## Milestones
This is a big milestone in my second technical challenge as I pivoted to a machine-learning based recommendation approach. 

## Resources
I used the TensorFlow and Keras documentation and online guide heavily in this PR. It really helped me understand what each line of code did and the arguments I could change to optimize my model. 
https://www.tensorflow.org/api_docs/python/tf/keras
https://www.tensorflow.org/guide/keras/

## Test Plan
I tested the model on many different parameters. I lessened the number of epochs to prevent overfitting, as well as added early stopping and reducing learning rate on plateau. I also lessened rare words (words with frequency < 5) to reduce the overall vocabulary and to have better generalization. I also realized I had to expand my number of articles so the model would have more content to train on. I increased my number of articles from 400 to over 1800. I also added shuffling of training data between epochs, as well as drop out to prevent the model from memorizing the training data.

### CODE REVIEW

**Review by nsclass — COMMENTED** [review]
Seem like that many interesting changes are made on vectorizer but it would be nice to have some organized way to explain what each Python module does though :) and not finished work yet? but it's understandable, may be due to not enough time.

File: python_scripts/recommendation_model/vectorizer/create_model.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    return embedding_matrix
+    
+def create_model(num_docs, word2int):
+    
+    # Model Parameters
+    embedding_dim = 100
+    vocab_size = len(word2int) + 1
+    num_docs = num_docs 
+
+    embedding_matrix = load_glove_embeddings('./glove.6B.100d.txt', word2int, embedding_dim)
```
[the remark this answers is not in the document]
Comment (chelsealiannegomez) [reply python_scripts/recommendation_model/vectorizer/create_model.py:85]: I forgot to mention this in the PR, but I used pre-trained embeddings as a basis for the embeddings of the words that are in the pre-trained set. I downloaded the GloVe set from here: https://nlp.stanford.edu/projects/glove/

**Review by rucharangnekar — APPROVED** [review]

**Comment by rucharangnekar** [comment]
looks good, but please try to put out smaller PRs for easier review and more incremental changes.

File: python_scripts/recommendation_model/vectorizer/save_doc_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+embedding_model = tf.keras.Model(inputs=model.input, outputs=doc_embedding_layer.output)
+
+doc_embeddings = embedding_model.predict([dummy_word_input, doc_indices])
+np.save('doc_embeddings.npy', doc_embeddings)
+np.savetxt('doc_embeddings.csv', doc_embeddings, delimiter=',')
+
+with open('word2int.json','r') as file:
+    word2int = json.load(file)
+
+query = "children children children children children children"
```
[the remark this answers is not in the document]
Comment (chelsealiannegomez) [reply python_scripts/recommendation_model/vectorizer/save_doc_embeddings.py:58]: Oops that was for testing, was just testing if increasing the number of times a word is in the query will increase keyword matching
